src/graph/graph.c

```diff
@@ -60,6 +66,86 @@ void _Graph_ResizeNodes(Graph *g, size_t n) {
+    /* As replacement gets a new ID, in case it has a Self edge,
+     * we need to update its entry. */
+    bool selfEdge = false;
+    GrB_Matrix_extractElement_BOOL(&selfEdge, M, replacementNodeID, replacementNodeID);
+    
+    if(selfEdge) {
+        GrB_Matrix_setElement_BOOL(M, true, replacementNodeID, removedNodeID);
+        GrB_Matrix_setElement_BOOL(M, false, replacementNodeID, replacementNodeID);
+    }
+
+    // Replace row.
+    GrB_Col_extract(row, NULL, NULL, M, GrB_ALL, nrows, replacementNodeID, desc);
+    GrB_Row_assign(M, NULL, NULL, row, removedNodeID, GrB_ALL, nrows, NULL);
+
+    // Replace col.
+    GrB_Col_extract(col, NULL, NULL, M, GrB_ALL, nrows, replacementNodeID, NULL);
+    GrB_Col_assign(M, NULL, NULL, col, GrB_ALL, nrows, removedNodeID, NULL);
+
+    // Clear replacemnt row and column.
+    GrB_Col_assign(M, NULL, NULL, zero, GrB_ALL, nrows, replacementNodeID, NULL);
+    GrB_Row_assign(M, NULL, NULL, zero, replacementNodeID, GrB_ALL, nrows, NULL);
+}
+
+/* Replace removed node ID relations with 
+ * its replacement node relations. */
+void _Graph_RemoveNodeRelations(Graph *g, int id, int replacementID) {    
+    GrB_Index nrows = g->node_cap;
+    
+    // Vectors to hold replacement relations.
+    GrB_Vector col; // Incoming edges.
+    GrB_Vector row; // Outgoing edges.
+    
+    GrB_Vector_new(&col, GrB_BOOL, nrows);
+    GrB_Vector_new(&row, GrB_BOOL, nrows);
+
+    // Descriptor used to transpose a matrix for row extraction.
+    GrB_Descriptor desc;
+    GrB_Descriptor_new(&desc);
+    GrB_Descriptor_set(desc, GrB_INP0, GrB_TRAN);
+
+    // Zero vector to clear entire row/column.
+    GrB_Vector zero;
+    GrB_Vector_new(&zero, GrB_BOOL, nrows);
+
+    /* Replace node ID relations with replacement ID relations within
+     * every relation matrix. */
+    _Graph_RemoveNodeFromMatrix(g->adjacency_matrix, id, replacementID, col, row, nrows, zero, desc);
+    for(int i = 0; i < g->relation_count; i++) {
+        _Graph_RemoveNodeFromMatrix(g->relations[i], id, replacementID, col, row, nrows, zero, desc);
+    }
+
+    // TODO: How many entries did we clear?, used later for GC.
+    // Clean up
+    GrB_Vector_free(&zero);
+    GrB_Vector_free(&col);
+    GrB_Vector_free(&row);
+    GrB_Descriptor_free(&desc);
+}
+
+/* Removes given node from graph's node blockchain,
+ * by replacing it with the node at the front of the chain. */
+void _Graph_RemoveNodeFromBlockchain(Graph *g, int id, int *replacementID) {
+    // Get the block in which removed node resides.
+    NodeBlock *removedNodeBlock = GRAPH_GET_NODE_BLOCK(g, id);
+    // Get node position within its block.
+    int nodeBlockIdx = GRAPH_NODE_POSITION_WITHIN_BLOCK(id);
+    // Get the newest node in the graph.
+    GrB_Index newestNodeID = g->node_count-1;
+    Node *newestNode = Graph_GetNode(g, newestNodeID);
+    // Replace removed node with newest node.
+    removedNodeBlock->nodes[nodeBlockIdx] = *newestNode;
```

Comment:
Won't this cause later attempts to access the newest node in the `NodeBlock` to fail? Its ID will still point to the latest block and offset.

Reply:
No, as IDs are internal, and are computed by the position of an entity within a matrix,
and so a node which was at row I and got moved to row J will have its ID update at some point to J,
we're updating node IDs lazily.